tests/federation_room_join_partial_state_test.go

```diff
@@ -563,6 +565,128 @@ func TestPartialStateJoin(t *testing.T) {
 			t.Errorf("Did not find derek's m.room.member event in gappy /sync response: %s", err)
 		}
 	})
+
+	// regression test for https://github.com/matrix-org/synapse/issues/13001
+	//
+	// There was an edge case where, if we initially receive lots of events as outliers,
+	// and they then get de-outliered as partial state events,
+	t.Run("Resync works with many prev_events with partial state", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+		psjResult := beginPartialStateJoin(t, deployment, alice)
+		defer psjResult.Destroy()
+
+		// Alice has now joined the room, and the server is syncing the state in the background.
+
+		// utility function to wait for a given event to arrive at the remote server.
+		// This works simply by polling /event until we get a 200.
+		awaitEventArrival := func(eventID string) {
+			start := time.Now()
+			for time.Since(start) < 10*time.Second {
+				res := alice.DoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", psjResult.ServerRoom.RoomID, "event", eventID})
+				eventResBody := client.ParseJSON(t, res)
+				if res.StatusCode == 200 {
+					t.Logf("Alice successfully received event %s", eventID)
+					return
+				}
+				if res.StatusCode == 404 && gjson.GetBytes(eventResBody, "errcode").String() == "M_NOT_FOUND" {
+					t.Logf("Fetching event %s failed with M_NOT_FOUND; will retry", eventID)
+					time.Sleep(1000 * time.Millisecond)
+					continue
+				}
+				t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
+			}
+			t.Fatalf("timeout waiting for event %s to be received", eventID)
+		}
```

Comment:
It was suggested that we use `WithRetryUntil` here: https://github.com/matrix-org/complement/pull/419#discussion_r928706407 , which looks like identical logic with a different timeout.

We might also want to make this helper available to other tests (and deduplicate the copy from #419)

Reply:
good idea, thanks. Done.